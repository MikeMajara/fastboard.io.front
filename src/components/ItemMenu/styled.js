import styled from '@xstyled/styled-components';
import { getSize } from '@xstyled/system';

export const SWrapper = styled.div`
  position: absolute;
  z-index: top;
  top: ${({ top }) => top}px;
  left: ${({ left }) => left}px;
  width: ${getSize('itemMenuWidth')};
  box-sizing: border-box;
  border-width: default;
  border-style: default;
  border-color: border;
  background-color: bg;
  border-radius: default;
  box-shadow: default;
  opacity: ${({ show }) => (show ? '1' : '0')};
  visibility: ${({ show }) => (show ? 'visible' : 'hidden')};
`;

export const SItem = styled.div`
  padding: xs md;

  transition: default;

  &:hover {
    background-color: bgHover;
    cursor: pointer;
  }
`;
