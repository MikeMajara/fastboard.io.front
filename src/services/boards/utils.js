/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
