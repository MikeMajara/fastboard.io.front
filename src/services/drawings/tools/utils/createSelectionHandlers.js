import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
