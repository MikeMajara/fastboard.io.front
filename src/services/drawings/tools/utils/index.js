export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
