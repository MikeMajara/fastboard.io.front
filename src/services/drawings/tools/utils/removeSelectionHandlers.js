import { Log } from '#utils';

/**
 * Remove selection handlers.
 *
 * @param {Object} selectedItem Item to be removed selection handlers from.
 * @param {Object} selectionLayer Layer that contains the selection handlers.
 */
export default function removeSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : removeSelectionHandlers');

  selectedItem.set({ selected : false });
  selectionLayer.removeChildren();
}
