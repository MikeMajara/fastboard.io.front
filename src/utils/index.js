export { default as Log } from './logger';
export { default as generateId } from './generateId';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as validBoardId } from './validBoardId';
