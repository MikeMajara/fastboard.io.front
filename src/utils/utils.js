/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
