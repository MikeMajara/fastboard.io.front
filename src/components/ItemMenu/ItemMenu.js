import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { SWrapper, SItem } from './styled';

// @todo icons for the menu items?
const ItemMenu = ({
  show, top, left, onItem2Front, onItem2Back, onItemRemove,
}) => {
  const { t } = useTranslation('board');

  return (
    <SWrapper show={show} top={top} left={left}>
      <SItem onClick={onItem2Front}>
        {t('itemMenu.bringForward')}
      </SItem>
      <SItem onClick={onItem2Back}>
        {t('itemMenu.sendBackward')}
      </SItem>
      <SItem onClick={onItemRemove}>
        {t('itemMenu.remove')}
      </SItem>
    </SWrapper>
  );
};

ItemMenu.defaultProps = {
  top  : 0,
  left : 0,
};

ItemMenu.propTypes = {
  show         : PropTypes.bool.isRequired,
  top          : PropTypes.number,
  left         : PropTypes.number,
  onItem2Front : PropTypes.func.isRequired,
  onItem2Back  : PropTypes.func.isRequired,
  onItemRemove : PropTypes.func.isRequired,
};

export default ItemMenu;
