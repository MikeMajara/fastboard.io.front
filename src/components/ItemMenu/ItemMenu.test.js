import React from 'react';
import { fireEvent } from '@testing-library/react';

import { render } from '#devTools/tests/utils';

import ItemMenu from './ItemMenu';

describe('Component : ItemMenu', () => {
  test('Is visible and reacts to actions', () => {
    const onItem2FrontFn = jest.fn();
    const onItem2BackFn = jest.fn();
    const onItemRemoveFn = jest.fn();
    const { getByText } = render(
      <ItemMenu
        show
        top={20}
        left={20}
        onItem2Front={onItem2FrontFn}
        onItem2Back={onItem2BackFn}
        onItemRemove={onItemRemoveFn}
      />,
    );

    expect(getByText('itemMenu.bringForward')).toBeInTheDocument();
    expect(getByText('itemMenu.sendBackward')).toBeInTheDocument();
    expect(getByText('itemMenu.remove')).toBeInTheDocument();

    fireEvent.click(getByText('itemMenu.bringForward'));
    fireEvent.click(getByText('itemMenu.sendBackward'));
    fireEvent.click(getByText('itemMenu.remove'));

    expect(onItem2FrontFn).toHaveBeenCalledTimes(1);
    expect(onItem2BackFn).toHaveBeenCalledTimes(1);
    expect(onItemRemoveFn).toHaveBeenCalledTimes(1);
  });

  test('Snapshot : visible', () => {
    const { asFragment } = render(
      <ItemMenu
        show
        top={20}
        left={20}
        onItem2Front={jest.fn()}
        onItem2Back={jest.fn()}
        onItemRemove={jest.fn()}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });

  test('Snapshot : not visible', () => {
    const { asFragment } = render(
      <ItemMenu
        show={false}
        top={20}
        left={20}
        onItem2Front={jest.fn()}
        onItem2Back={jest.fn()}
        onItemRemove={jest.fn()}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
