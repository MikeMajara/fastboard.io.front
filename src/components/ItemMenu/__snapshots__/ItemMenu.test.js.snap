// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : ItemMenu Snapshot : not visible 1`] = `
<DocumentFragment>
  <div
    class="sc-pscky hrVfUg"
  >
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.bringForward
    </div>
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.sendBackward
    </div>
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.remove
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component : ItemMenu Snapshot : visible 1`] = `
<DocumentFragment>
  <div
    class="sc-pscky jlucIJ"
  >
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.bringForward
    </div>
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.sendBackward
    </div>
    <div
      class="sc-qQYBZ kyFabS"
    >
      itemMenu.remove
    </div>
  </div>
</DocumentFragment>
`;
